### ISSUE TITLE
Support /secutiry API of Artifactory

### ISSUE THREAD

**Issue by g4s8** [artipie/artipie#444]
Support [Artifactory API](https://www.jfrog.com/confluence/display/rtf/artifactory+rest+api) at `/api/secutiry/users` and `/api/secutiry/permissions`: API clients should be able to create update and delete users and permissions.

**Comment by olenagerasimova** [comment]
@g4s8 are we are going to support all requests to `/api/secutiry/users` endpoint? There are several variations:
1. GET `/api/security/users` to obtain users list
2. GET `/api/security/users/{userName}` to obtain user info
3. PUT `/api/security/users/{userName} ` to create/replace user
4. POST `/api/security/users/{userName}` to update user
5. DELETE `/api/security/users/{userName}` to delete user
4. GET `/api/security/encryptedPassword` Get User Encrypted Password
6. POST `/api/security/users/authorization/changePassword` to change password
7. Several methods to expire/unexpire user password, expire password for user list, expire passwords for all users

2-4 work with json body [format](https://www.jfrog.com/confluence/display/JFROG/Security+Configuration+JSON#SecurityConfigurationJSON-application/vnd.org.jfrog.artifactory.security.User+json) where `email` field is required, we do not store user's emails.

**Comment by g4s8** [comment]
@olenagerasimova we can start with CRUD operations for users (`GET /api/security/users`, `GET /api/security/users/{userName}`, `PUT /api/security/users/{userName}`, `POST /api/security/users/{userName}` and `DELETE /api/security/users/{userName}`). All other `/users` operations seems to be not necessary.

**Comment by olenagerasimova** [comment]
@g4s8 ok, please check user artifactory [payload](https://www.jfrog.com/confluence/display/JFROG/Security+Configuration+JSON) 
```
{
  - "name": "davids",
  + "email" : "davids@jfrog.com",
  + "password": "***" (write-only, never returned),
  - "admin": false (default),
  - "profileUpdatable": true (default),
  - "disableUIAccess" : true,
  - "internalPasswordDisabled": false (default),
  ! "lastLoggedIn": ISO8601 (yyyy-MM-dd'T'HH:mm:ss.SSSZ),
  ! "realm": "Internal",
  - "groups" : [ "deployers", "users" ],
  - "watchManager": false(default),
  - "policyManager": false(default)
}
``` 
fields with + are required, - optional, ! read only. Questions:
1. We do not have users' email to return - what should I do about it?
2. What exactly should be done on user creation? Obviously, I should add corresponding record to `credentials.yaml`, what about repository permissions?
3. What should I return in `lastLoggedIn` and `realm` fields?
4. Do I understand correctly that we simply ignore fields we cannot process properly (all except name and password)?

**Comment by olenagerasimova** [comment]
@g4s8 `/permission` also have several endpoints:
1. GET `/api/security/permissions` - Get the permission targets list, what is permission target in our terms?
2. GET `/api/security/permissions/{permissionTargetName}` - Get the details of an Artifactory Permission Target, returns json [payload](https://www.jfrog.com/confluence/display/JFROG/Security+Configuration+JSON#SecurityConfigurationJSON-application/vnd.org.jfrog.artifactory.security.PermissionTarget+json)
3. PUT `/api/security/permissions/{permissionTargetName}` - Creates/replaces a new permission target, accepts [json body](https://www.jfrog.com/confluence/display/JFROG/Security+Configuration+JSON#SecurityConfigurationJSON-application/vnd.org.jfrog.artifactory.security.PermissionTargetV2+json)
4. GET `/api/storage/{repoKey}/{itemPath}?permissions` - Returns a list of effective permissions for the specified item 

Are we going to implement all these methods and how are we going to process requests?

**Comment by g4s8** [comment]
@olenagerasimova I've got some answers for you:

> We do not have users' email to return - what should I do about it?

We can ignore this field and return any data. I'd use `{username}@artipie.com` or similar

> What exactly should be done on user creation? Obviously, I should add corresponding record to credentials.yaml, what about repository permissions?

We need to create the user only, permissions for repository are managed via `/permissions` API

> hat should I return in lastLoggedIn and realm fields?

Any valid data, it doesn't matter

> Do I understand correctly that we simply ignore fields we cannot process properly (all except name and password)?

We can ignore password changing for first release, but we need to support it later: it should be possible to change the passord via API

**Comment by olenagerasimova** [comment]
@g4s8 one more question: `/api/security/users` returns the following payload:
```
GET /api/security/users
[
  {
    "name": "davids"
    "uri" : "http://localhost:8081/artifactory/api/security/users/davids"
    "realm" : "internal"
  }, {
    "name": "danl"
    "uri" : "http://localhost:8081/artifactory/api/security/users/danl"
    "realm" : "ldap"
  }
]
```
where from host and port can be obtained?

**Comment by g4s8** [comment]
> @g4s8 one more question: `/api/security/users` returns the following payload:
> ```
> GET /api/security/users
> [
>   {
>     "name": "davids"
>     "uri" : "http://localhost:8081/artifactory/api/security/users/davids"
>     "realm" : "internal"
>   }, {
>     "name": "danl"
>     "uri" : "http://localhost:8081/artifactory/api/security/users/danl"
>     "realm" : "ldap"
>   }
> ]
> ```
> where from host and port can be obtained?

@olenagerasimova most probably we need configuration parameters in `artipie.yml` for base URI

**Comment by jeffzhu503** [comment]
What is the data model and data storage behind this API?  Is the user at the system level or repository level?

**Comment by olenagerasimova** [comment]
@jeffzhu503 this api will cover Artipie credentials and repository permissions settings. User is configured on the whole Artipie level, permissions for repository can be set for Artipie users. User information (for now we can work with username and password only) is stored as yaml file (this file location is configured), repository permissions are configured in repository config.

**Comment by jeffzhu503** [comment]
>  User information (for now we can work with username and password only) is stored as yaml file (this file location is configured), repository permissions are configured in repository config.

What is the name of the file?  What is the format of these yaml file? Is this a single file which might save thousands of users and thousands of repositories? How do you manage the relationship of users and repositories?  How do you control the write/read of this file? How do you prevent the lock or race condition of IO in this file? How do you scale the read/write of this files when the number of users increase exponentially?

**Comment by 0pdd** [comment]
@g4s8 the puzzle [#464](https://github.com/artipie/artipie/issues/464) is still not solved.

**Comment by olenagerasimova** [comment]
@jeffzhu503 please check https://github.com/artipie/artipie/issues/146 (with comments)- there you find all the information about users and credentials. Repository permissions are configured for each repository in repository config, please, check https://github.com/artipie/artipie/issues/69

**Comment by 0pdd** [comment]
@g4s8 2 puzzles [#464](https://github.com/artipie/artipie/issues/464), [#472](https://github.com/artipie/artipie/issues/472) are still not solved.

**Comment by 0pdd** [comment]
@g4s8 4 puzzles [#464](https://github.com/artipie/artipie/issues/464), [#472](https://github.com/artipie/artipie/issues/472), [#479](https://github.com/artipie/artipie/issues/479), [#480](https://github.com/artipie/artipie/issues/480) are still not solved.

**Comment by 0pdd** [comment]
@g4s8 2 puzzles [#464](https://github.com/artipie/artipie/issues/464), [#484](https://github.com/artipie/artipie/issues/484) are still not solved; solved: [#472](https://github.com/artipie/artipie/issues/472), [#479](https://github.com/artipie/artipie/issues/479), [#480](https://github.com/artipie/artipie/issues/480).

**Comment by 0pdd** [comment]
@g4s8 6 puzzles [#464](https://github.com/artipie/artipie/issues/464), [#484](https://github.com/artipie/artipie/issues/484), [#494](https://github.com/artipie/artipie/issues/494), [#495](https://github.com/artipie/artipie/issues/495), [#496](https://github.com/artipie/artipie/issues/496), [#497](https://github.com/artipie/artipie/issues/497) are still not solved; solved: [#472](https://github.com/artipie/artipie/issues/472), [#479](https://github.com/artipie/artipie/issues/479), [#480](https://github.com/artipie/artipie/issues/480).

**Comment by 0pdd** [comment]
@g4s8 5 puzzles [#464](https://github.com/artipie/artipie/issues/464), [#484](https://github.com/artipie/artipie/issues/484), [#494](https://github.com/artipie/artipie/issues/494), [#495](https://github.com/artipie/artipie/issues/495), [#497](https://github.com/artipie/artipie/issues/497) are still not solved; solved: [#472](https://github.com/artipie/artipie/issues/472), [#479](https://github.com/artipie/artipie/issues/479), [#480](https://github.com/artipie/artipie/issues/480), [#496](https://github.com/artipie/artipie/issues/496).

**Comment by 0pdd** [comment]
@g4s8 4 puzzles [#464](https://github.com/artipie/artipie/issues/464), [#494](https://github.com/artipie/artipie/issues/494), [#495](https://github.com/artipie/artipie/issues/495), [#497](https://github.com/artipie/artipie/issues/497) are still not solved; solved: [#472](https://github.com/artipie/artipie/issues/472), [#479](https://github.com/artipie/artipie/issues/479), [#480](https://github.com/artipie/artipie/issues/480), [#484](https://github.com/artipie/artipie/issues/484), [#496](https://github.com/artipie/artipie/issues/496).

**Comment by 0pdd** [comment]
@g4s8 4 puzzles [#464](https://github.com/artipie/artipie/issues/464), [#494](https://github.com/artipie/artipie/issues/494), [#497](https://github.com/artipie/artipie/issues/497), [#509](https://github.com/artipie/artipie/issues/509) are still not solved; solved: [#472](https://github.com/artipie/artipie/issues/472), [#479](https://github.com/artipie/artipie/issues/479), [#480](https://github.com/artipie/artipie/issues/480), [#484](https://github.com/artipie/artipie/issues/484), [#495](https://github.com/artipie/artipie/issues/495), [#496](https://github.com/artipie/artipie/issues/496).

**Comment by 0pdd** [comment]
@g4s8 3 puzzles [#464](https://github.com/artipie/artipie/issues/464), [#494](https://github.com/artipie/artipie/issues/494), [#509](https://github.com/artipie/artipie/issues/509) are still not solved; solved: [#472](https://github.com/artipie/artipie/issues/472), [#479](https://github.com/artipie/artipie/issues/479), [#480](https://github.com/artipie/artipie/issues/480), [#484](https://github.com/artipie/artipie/issues/484), [#495](https://github.com/artipie/artipie/issues/495), [#496](https://github.com/artipie/artipie/issues/496), [#497](https://github.com/artipie/artipie/issues/497).

**Comment by 0pdd** [comment]
@g4s8 2 puzzles [#464](https://github.com/artipie/artipie/issues/464), [#509](https://github.com/artipie/artipie/issues/509) are still not solved; solved: [#472](https://github.com/artipie/artipie/issues/472), [#479](https://github.com/artipie/artipie/issues/479), [#480](https://github.com/artipie/artipie/issues/480), [#484](https://github.com/artipie/artipie/issues/484), [#494](https://github.com/artipie/artipie/issues/494), [#495](https://github.com/artipie/artipie/issues/495), [#496](https://github.com/artipie/artipie/issues/496), [#497](https://github.com/artipie/artipie/issues/497).

**Comment by g4s8** [comment]
@olenagerasimova thanks

**Comment by 0pdd** [comment]
@g4s8 the puzzle [#464](https://github.com/artipie/artipie/issues/464) is still not solved; solved: [#472](https://github.com/artipie/artipie/issues/472), [#479](https://github.com/artipie/artipie/issues/479), [#480](https://github.com/artipie/artipie/issues/480), [#484](https://github.com/artipie/artipie/issues/484), [#494](https://github.com/artipie/artipie/issues/494), [#495](https://github.com/artipie/artipie/issues/495), [#496](https://github.com/artipie/artipie/issues/496), [#497](https://github.com/artipie/artipie/issues/497), [#509](https://github.com/artipie/artipie/issues/509).

**Comment by 0pdd** [comment]
@g4s8 all 10 puzzles are solved here: [#464](https://github.com/artipie/artipie/issues/464), [#472](https://github.com/artipie/artipie/issues/472), [#479](https://github.com/artipie/artipie/issues/479), [#480](https://github.com/artipie/artipie/issues/480), [#484](https://github.com/artipie/artipie/issues/484), [#494](https://github.com/artipie/artipie/issues/494), [#495](https://github.com/artipie/artipie/issues/495), [#496](https://github.com/artipie/artipie/issues/496), [#497](https://github.com/artipie/artipie/issues/497), [#509](https://github.com/artipie/artipie/issues/509).
